components/script/dom/shadowroot.rs

```diff
@@ -70,6 +77,16 @@ impl ShadowRoot {
         )
     }
 
+    pub fn detach(&self) {
+        self.host.set(None);
+        let node = self.upcast::<Node>();
+        node.set_containing_shadow_root(None);
```

Comment:
This is still not enough, this doesn't clear the containing shadow of descendants or anything like that. Can we call RemoveChild on the children so that the tree ends up torn into pieces? That'd make me more comfortable since then the only state you need to keep an eye on is the ShadowRoot state.